### PR TITLE
`data_collection.rs` stability improvements

### LINKED ISSUE #590: test: `cargo test data_output` fails
# Problem
This test fails a about ⅓ of the time we run it. We would like to figure out why and get it to pass so that the expected behavior is correct. 

I think this is one of the last remaining tasks for the feature branch.

### PR DESCRIPTION
**Pull request by kinrezC**
- Refactors `data_collection.rs` to avoid any race conditions. This PR makes the whole event logger process sync and adds a signal from the environment to indicate when the process should shut down.
- Default data output is now json
- Future improvements should involve adding new data output formats
- Closes #590

### CODE REVIEW

File: arbiter-core/src/data_collection.rs [hunk 1]
```diff
@@ -184,12 +162,70 @@ impl EventLogger {
     /// This function will return an error if there is a problem creating the
     /// directories or files, or writing to the files.
     pub fn run(self) -> Result<(), RevmMiddlewareError> {
-        tokio::spawn(async move {
-            let mut set = self.events;
-            while let Some(res) = set.join_next().await {
-                info!("task completed: {:?}", res);
+        let receiver = self.receiver.unwrap();
+        let dir = self.directory.unwrap_or("./out".into());
```
Comment (Alexangelj) [inline arbiter-core/src/data_collection.rs:166]: Does this overwrite if an existing directory is there? Might conflict with default foundry output which is `out`, if it does an it overwrites the foundry out that could make issues

File: arbiter-core/src/environment/mod.rs [hunk 2]
```diff
@@ -668,9 +685,19 @@ impl EventBroadcaster {
...
-            sender.send(logs.clone())?;
+    fn broadcast(&self, logs: Option<Vec<Log>>, stop_signal: bool) -> Result<(), EnvironmentError> {
+        if stop_signal {
+            for sender in &self.0 {
+                sender.send(Broadcast::StopSignal)?;
+            }
+            return Ok(());
+        } else {
+            if logs.is_none() {
+                panic!("This shouldn't happen, but we should probably make this an error. Somehow we got told to broadcast `None` logs!");
```
Comment (Alexangelj) [inline arbiter-core/src/environment/mod.rs:696]: lol is this possible to hit?

File: arbiter-core/src/tests/data_output.rs [hunk 3]
```diff
@@ -38,16 +38,17 @@ async fn data_capture() {
...
     }
-    tokio::fs::remove_dir_all("./test_output1").await.unwrap();
+
+    env.stop();
+    tokio::fs::remove_dir_all("./out").await.unwrap();
 }
 
 #[traced_test]
 #[tokio::test(flavor = "multi_thread")]
 async fn data_capture_output_validation() {
```
Comment (Alexangelj) [inline arbiter-core/src/tests/data_output.rs:48]: good test 🙏

File: arbiter-core/src/tests/output_test.json [hunk 4]
```diff
@@ -0,0 +1 @@
...0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","spender":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","amount":"0x1"},{"owner":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","spender":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","amount":"0x1"},{"owner":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","spender":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","amount":"0x1"}]},"arby":{"ApprovalFilter":[{"owner":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","spender":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","amount":"0x1"},{"owner":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","spender":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","amount":"0x1"},{"owner":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","spender":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","amount":"0x1"},{"owner":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","spender":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","amount":"0x1"},{"owner":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","spender":"0x2efdc9eecfee3a776209fcb8e9a83a6b221d74f5","amount":"0x1"}]},"lex":{"PriceChangeFilter":[{"price":"0xde0b6b3a7640000"},{"price":"0xde0b6b3a7640000"},{"price":"0xde0b6b3a7640000"},{"price":"0xde0b6b3a7640000"},{"price":"0xde0b6b3a7640000"}]}}
```
Comment (Alexangelj) [inline arbiter-core/src/tests/output_test.json:1]: okay with the labels is nice

File: arbiter-core/src/data_collection.rs [hunk 5]
```diff
@@ -73,93 +85,59 @@ impl EventLogger {
...
-                    file.write_all(values.as_bytes()).await.unwrap();
-                    file.write_all("\n".as_bytes()).await.unwrap();
-                }
-                continue;
-            }
-        });
+        // Grab the connection from the client and add a new event sender so that we
+        // have a distinct channel to now receive events over
+        let event_transmuted: EventTransmuted<Arc<RevmMiddleware>, RevmMiddleware, D> =
+            unsafe { transmute(event) };
```
Comment (Alexangelj) [inline arbiter-core/src/data_collection.rs:96]: idk what this does or why its here, maybe a comment?

**Review by Alexangelj — APPROVED** [review]
Let me know if I am getting this right on whats going on here now:
- We made a new dedicated broadcast channel
- We have that channel run on a separate normal thread
- We push each event through this channel and store it in the same place so its output to the same file

File: arbiter-core/src/data_collection.rs (hunk 5 above)
Comment (0xJepsen) [reply arbiter-core/src/data_collection.rs:96]: Same, is the kinda same unsafe stuff we did with resolving pending transactions?

**Review by 0xJepsen — APPROVED** [review]
Remarkable work on this, you guys are beasts.

File: arbiter-core/src/data_collection.rs (hunk 5 above)
Comment (kinrezC) [reply arbiter-core/src/data_collection.rs:96]: lets you access private fields

File: arbiter-core/src/data_collection.rs (hunk 1 above)
Comment (Autoparallel) [reply arbiter-core/src/data_collection.rs:166]: Good catch. @kinrezC let's change this do something like `./data_output`

File: arbiter-core/src/environment/mod.rs (hunk 2 above)
Comment (Autoparallel) [reply arbiter-core/src/environment/mod.rs:696]: I really don't think so. I think if no logs, you actually broadcast `vec![]`. That could actually be optimized out...

File: arbiter-core/src/tests/output_test.json (hunk 4 above)
Comment (Autoparallel) [reply arbiter-core/src/tests/output_test.json:1]: And unraveling it is super clean!

File: arbiter-core/src/data_collection.rs (hunk 5 above)
Comment (Autoparallel) [reply arbiter-core/src/data_collection.rs:96]: @Alexangelj if you're talking about the transmute, it is essentially what Matt said.

So you can take an `Event<Arc<RevmMiddleware>, RevmMiddleware, D>` and cast it into the `EventTransmuted<Arc<RevmMiddleware, RevmMiddleware, D>` which then gives us direct access to the fields of the struct.

**Pull request merged**